src/per/uart.cpp

```diff
@@ -215,106 +481,409 @@ void HAL_UART_AbortReceiveCpltCallback(UART_HandleTypeDef* huart)
+
+    GPIO_TypeDef* ports[] = {GPIOA,
+                             GPIOB,
+                             GPIOC,
+                             GPIOD,
+                             GPIOE,
+                             GPIOF,
+                             GPIOG,
+                             GPIOH,
+                             GPIOI,
+                             GPIOJ,
+                             GPIOK};
+
+    for(int i = 0; i < 11; i++)
+    {
+        if(port == ports[i])
+        {
+            func_p[i]();
+            return;
+        }
+    }
+}
+
+//more silly macro expansion
+void USART1_CLK_ENABLE()
+{
+    __HAL_RCC_USART1_CLK_ENABLE();
+};
+void USART2_CLK_ENABLE()
+{
+    __HAL_RCC_USART2_CLK_ENABLE();
+};
+void USART3_CLK_ENABLE()
+{
+    __HAL_RCC_USART3_CLK_ENABLE();
+};
+void UART4_CLK_ENABLE()
+{
+    __HAL_RCC_UART4_CLK_ENABLE();
+};
+void UART5_CLK_ENABLE()
+{
+    __HAL_RCC_UART5_CLK_ENABLE();
+};
+void USART6_CLK_ENABLE()
+{
+    __HAL_RCC_USART6_CLK_ENABLE();
+};
+void UART7_CLK_ENABLE()
+{
+    __HAL_RCC_UART7_CLK_ENABLE();
+};
+void UART8_CLK_ENABLE()
+{
+    __HAL_RCC_UART8_CLK_ENABLE();
+};
+void LPUART1_CLK_ENABLE()
+{
+    __HAL_RCC_LPUART1_CLK_ENABLE();
+};
+
+void UartClockEnable(USART_TypeDef* periph)
+{
+    VoidFunc       func_p[] = {USART1_CLK_ENABLE,
+                         USART2_CLK_ENABLE,
+                         USART3_CLK_ENABLE,
+                         UART4_CLK_ENABLE,
+                         UART5_CLK_ENABLE,
+                         USART6_CLK_ENABLE,
+                         UART7_CLK_ENABLE,
+                         UART8_CLK_ENABLE,
+                         LPUART1_CLK_ENABLE};
+    USART_TypeDef* periphs[]
+        = {USART1, USART2, USART3, UART4, UART5, USART6, UART7, UART8, LPUART1};
+
+    for(int i = 0; i < 9; i++)
+    {
+        if(periph == periphs[i])
+        {
+            func_p[i]();
+            return;
+        }
+    }
+}
+
+void EnableNvic(USART_TypeDef* periph)
+{
+    IRQn_Type types[] = {USART1_IRQn,
+                         USART2_IRQn,
+                         USART3_IRQn,
+                         UART4_IRQn,
+                         UART5_IRQn,
+                         USART6_IRQn,
+                         UART7_IRQn,
+                         UART8_IRQn,
+                         LPUART1_IRQn};
+
+    USART_TypeDef* periphs[]
+        = {USART1, USART2, USART3, UART4, UART5, USART6, UART7, UART8, LPUART1};
+
+    for(int i = 0; i < 9; i++)
+    {
+        if(periphs[i] == periph)
+        {
+            HAL_NVIC_SetPriority(types[i], 0, 0);
+            HAL_NVIC_EnableIRQ(types[i]);
+            return;
+        }
+    }
+}
 
 // HAL Interface functions
 void HAL_UART_MspInit(UART_HandleTypeDef* uartHandle)
 {
-    GPIO_InitTypeDef GPIO_InitStruct = {0};
-    if(uartHandle->Instance == USART1)
+    UartHandler::Impl* handle = MapInstanceToHandle(uartHandle->Instance);
+    GpioClockEnable(handle->port);
+    UartClockEnable(handle->periph);
```

Comment:
There's no 'reuse' of the UartClockEnable. So I almost think a simple if/else or switch over the peripheral instance within this function would be a bit simpler.  Same goes for the NvicEnable. 

The wrappers for each clock enable/disable is pretty long. And the only time those will need to get called is in the MspInit/DeInit functions.

The other aspect as that these aren't class functions so if we decide to add any flexibility to the interrupt priority, or even whether an interrupt gets enabled through the Config struct there would be no way of passing that along without more drastic changes.